compiler/src/dotty/tools/dotc/typer/Implicits.scala

```diff
@@ -696,22 +697,45 @@ trait Implicits { self: Typer =>
         }
     }
     def location(preposition: String) = if (where.isEmpty) "" else s" $preposition $where"
+    def userDefinedMessage(annot: Annotation, params: List[String], args: List[Type]): Option[String] =
+      for (Trees.Literal(Constant(raw: String)) <- annot.argument(0)) yield {
+        err.userDefinedErrorString(
+          raw,
+          params,
+          args)
+      }
     arg.tpe match {
       case ambi: AmbiguousImplicits =>
-        msg(s"ambiguous implicit arguments: ${ambi.explanation}${location("of")}")(
-            s"ambiguous implicit arguments of type ${pt.show} found${location("for")}")
-      case _ =>
-        val userDefined =
-          for {
-            notFound <- pt.typeSymbol.getAnnotation(defn.ImplicitNotFoundAnnot)
-            Trees.Literal(Constant(raw: String)) <- notFound.argument(0)
+        val maybeAnnot = ambi.alt1.ref.symbol.getAnnotation(defn.ImplicitAmbiguousAnnot).map(
+          (_, ambi.alt1)
+        ).orElse(
+          ambi.alt2.ref.symbol.getAnnotation(defn.ImplicitAmbiguousAnnot).map(
+            (_, ambi.alt2)
+          )
+        )
```

Comment:
I would do it the other way around. First look for a user defined annotation message and then try try to recover type params only if needed:
```scala
def userDefinedMsg(sym: Symbol, cls: Symbol) = for {
  ann <- symbol.getAnnotation(cls)
  Trees.Literal(Constant(msg: String)) <- ann.argument(0)
} yield msg

object AmbiguousImplicitMsg {
  def unapply(search: SearchSuccess) =
    userDefinedMsg(search.ref.symbol, defn.ImplicitAmbiguousAnnot)
}

(ambi.alt1, ambi.alt2) match {
  case (AmbiguousImplicitMsg(msg), _) => ... // recover type params from alt1
  case (_, AmbiguousImplicitMsg(msg)) => ... // recover type params from alt2
  case _ =>
    msg(s"ambiguous implicit arguments: ${ambi.explanation}${location("of")}")(
        s"ambiguous implicit arguments of type ${pt.show} found${location("for")}")
}
```